File: common/src/main/java/cn/breadnicecat/candycraftce/block/blocks/CandyCropBlock.java
package cn.breadnicecat.candycraftce.block.blocks;

import cn.breadnicecat.candycraftce.block.CBlocks;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.Mth;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.IntegerProperty;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.NotNull;

import static java.lang.Math.min;

/**
 * @author <a href="https://gitee.com/Bread_NiceCat">Bread_NiceCat</a>
 * @date 2023/1/20 14:29
 */
public class CandyCropBlock extends CandyPlantBlock implements ISugarTarget {

	public static final int MAX_AGE = 7;
	public static final IntegerProperty AGE = BlockStateProperties.AGE_7;
	protected final VoxelShape[] shapes;
	private static VoxelShape[] SHAPE_L4 = new VoxelShape[]{
			box(0.0D, 0.0D, 0.0D, 16.0D, 2.0D, 16.0D),//0
			box(0.0D, 0.0D, 0.0D, 16.0D, 4.0D, 16.0D),//1
			box(0.0D, 0.0D, 0.0D, 16.0D, 8.0D, 16.0D),//2
			box(0.0D, 0.0D, 0.0D, 16.0D, 8.0D, 16.0D),//3
	};


	public int getMaxAge() {
		return MAX_AGE;
	}

	public CandyCropBlock(Properties properties, VoxelShape[] shapes) {
		super(properties);
		this.shapes = shapes;
		registerDefaultState(stateDefinition.any().setValue(AGE, 0));
	}

	public static CandyCropBlock createL4(Properties properties) {
		return new CandyCropBlock(properties, SHAPE_L4);
	}

	public int getAge(BlockState b) {
		return b.getValue(AGE);
	}


	/**
	 * @return [0, {@code shapes.length}]
	 */
	public int getStage(BlockState b) {
		return switch (getAge(b)) {
			case 0, 1, 2 -> 0;
			case 3, 4 -> 1;
			case 5, 6 -> 2;
			default -> 3;
		};
	}


	@SuppressWarnings("deprecation")
	@Override
	public @NotNull VoxelShape getShape(@NotNull BlockState pState, @NotNull BlockGetter pLevel, @NotNull BlockPos pPos, @NotNull CollisionContext pContext) {
		return shapes[getStage(pState)];
	}

	public boolean canSurvive(@NotNull BlockState pState, @NotNull LevelReader pLevel, @NotNull BlockPos pPos) {
		return pLevel.getBlockState(pPos.below()).is(CBlocks.PUDDING_FARMLAND.get())
				&& ((pLevel.getRawBrightness(pPos, 0) >= 8 || pLevel.canSeeSky(pPos)) && super.canSurvive(pState, pLevel, pPos));
	}


	//[VanillaCopy]net.minecraft.world.level.block.CropBlock
	@SuppressWarnings("deprecation")
	@Override
	public void randomTick(BlockState state, @NotNull ServerLevel level, BlockPos pos, RandomSource random) {
		int age;
		if (level.getRawBrightness(pos, 0) >= 9 && (age = this.getAge(state)) < this.getMaxAge() && random.nextInt((int) (25.0f / (getGrowthSpeed(this, level, pos))) + 1) == 0) {
			level.setBlock(pos, state.setValue(AGE, age + 1), 2);
		}
	}

	//[VanillaCopy]net.minecraft.world.level.block.CropBlock
	protected float getGrowthSpeed(Block block, BlockGetter level, @NotNull BlockPos pos) {
		boolean bl2;
		float f = 1.0f;
		BlockPos blockPos = pos.below();
		for (int i = -1; i <= 1; ++i) {
			for (int j = -1; j <= 1; ++j) {
				float g = 0.0f;
				BlockState blockState = level.getBlockState(blockPos.offset(i, 0, j));
				if (blockState.is(CBlocks.PUDDING_FARMLAND.get())) {
					g = 1.0f;
					if (blockState.getValue(PuddingFarmBlock.MOISTURE) > 0) {
						g = 3.0f;
					}
				}
				if (i != 0 || j != 0) {
					g /= 4.0f;
				}
				f += g;
			}
		}
		BlockPos blockPos2 = pos.north();
		BlockPos blockPos3 = pos.south();
		BlockPos blockPos4 = pos.west();
		BlockPos blockPos5 = pos.east();
		boolean bl = level.getBlockState(blockPos4).is(block) || level.getBlockState(blockPos5).is(block);
		boolean bl3 = bl2 = level.getBlockState(blockPos2).is(block) || level.getBlockState(blockPos3).is(block);
		if (bl && bl2) {
			f /= 2.0f;
		} else {
			boolean bl32;
			boolean bl4 = bl32 = level.getBlockState(blockPos4.north()).is(block) || level.getBlockState(blockPos5.north()).is(block) || level.getBlockState(blockPos5.south()).is(block) || level.getBlockState(blockPos4.south()).is(block);
			if (bl32) {
				f /= 2.0f;
			}
		}
		return f;
	}

	@Override
	public boolean isRandomlyTicking(@NotNull BlockState pState) {
		return getAge(pState) < getMaxAge();
	}

	@Override
	protected void createBlockStateDefinition(StateDefinition.@NotNull Builder<Block, BlockState> pBuilder) {
		pBuilder.add(AGE);
	}

	@Override
	public boolean isValidSugarTarget(LevelReader level, BlockPos pos, BlockState state, boolean isClient) {
		return getAge(state) < getMaxAge();
	}

	@Override
	public boolean isSugarSuccess(ServerLevel level, RandomSource rand, BlockPos pos, BlockState state) {
		return rand.nextFloat() < 0.80;
	}

	@Override
	public void performSugar(ServerLevel level, RandomSource rand, BlockPos pos, BlockState state) {
		int age = min(getAge(state) + Mth.nextInt(level.random, 2, 5), getMaxAge());
		level.setBlock(pos, state.setValue(AGE, age), 2);
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/core/block/blocks/plant/CandyCropBlock.kt
package cn.breadnicecat.candycraftce.core.block.blocks.plant

import cn.breadnicecat.candycraftce.core.block.CBlocks
import cn.breadnicecat.candycraftce.core.block.blocks.PuddingFarmBlock
import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.RandomSource
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraft.world.level.block.state.properties.IntegerProperty
import net.minecraft.world.phys.shapes.CollisionContext
import net.minecraft.world.phys.shapes.VoxelShape
import kotlin.math.min
import kotlin.random.Random

/**
 * Created by NiceCat on 2025/12/5.
 * Project: candycraftce
 * <p>
 *
 */
class CandyCropBlock(
    properties: Properties,
    val stages: (Int) -> Int,
    val shapes: (Int) -> VoxelShape,
) : CandyPlantBlock(properties), ISugarTarget {
    companion object {
        const val MAX_AGE: Int = 7
        val AGE: IntegerProperty = BlockStateProperties.AGE_7
        val SHAPE_L4 = arrayOf<VoxelShape>(
            box(0.0, 0.0, 0.0, 16.0, 2.0, 16.0),  //0
            box(0.0, 0.0, 0.0, 16.0, 4.0, 16.0),  //1
            box(0.0, 0.0, 0.0, 16.0, 8.0, 16.0),  //2
            box(0.0, 0.0, 0.0, 16.0, 8.0, 16.0),  //3
        )

        fun stagesL4(age: Int): Int = when (age) {
            0, 1, 2 -> 0
            3, 4 -> 1
            5, 6 -> 2
            else -> 3
        }

        fun shapesL4(stage: Int): VoxelShape = SHAPE_L4[stage]

        fun createL4(properties: Properties): CandyCropBlock {
            return CandyCropBlock(properties, ::stagesL4, ::shapesL4)
        }

    }

    fun getAge(b: BlockState): Int {
        return b.getValue(AGE)
    }

    fun getStage(b: BlockState): Int {
        return stages(getAge(b))
    }

    @Deprecated("Deprecated in Java")
    override fun getShape(
        state: BlockState,
        level: BlockGetter,
        pos: BlockPos,
        context: CollisionContext,
    ): VoxelShape = shapes(getStage(state))

    @Suppress("DEPRECATION")
    @Deprecated("Deprecated in Java")
    override fun canSurvive(state: BlockState, level: LevelReader, pos: BlockPos): Boolean {
        return level.getBlockState(pos.below()).`is`(CBlocks.sugar_block.block)
                && (level.canSeeSky(pos) || level.getRawBrightness(pos, 0) >= 8)
                && super.canSurvive(state, level, pos)
    }

    /**
     * VanillaCopy
     * @see net.minecraft.world.level.block.CropBlock.randomTick
     */
    @Deprecated("Deprecated in Java")
    override fun randomTick(state: BlockState, level: ServerLevel, pos: BlockPos, random: RandomSource) {
        val age: Int = getAge(state)
        if (level.getRawBrightness(pos, 0) >= 9
            && age < MAX_AGE
            && Random.nextInt((25f / getGrowthSpeed(this, level, pos)).toInt() + 1) == 0
        ) {
            level.setBlock(pos, state.setValue(AGE, age + 1), 2)
        }
    }

    /** VanillaCopy
     *  change `farm` to `pudding_farm`
     *  @see net.minecraft.world.level.block.CropBlock.getGrowthSpeed
     */
    fun getGrowthSpeed(block: Block, level: BlockGetter, pos: BlockPos): Float {
        var speed = 1.0f
        val blockPos = pos.below()
        for (x in -1..1) {
            for (z in -1..1) {
                var boost = 0.0f
                val blockState = level.getBlockState(blockPos.offset(x, 0, z))
                if (blockState.`is`(CBlocks.pudding_farm.block)) {
                    boost = 1.0f
                    if (blockState.getValue(PuddingFarmBlock.MOISTURE) > 0) {
                        boost = 3.0f
                    }
                }
                if (x != 0 || z != 0) {
                    boost /= 4.0f
                }
                speed += boost
            }
        }
        val north = pos.north()
        val south = pos.south()
        val west = pos.west()
        val east = pos.east()

        //东或西 南或北 同时种植相同的作物
        if ((level.getBlockState(west).`is`(block)
                    || level.getBlockState(east).`is`(block))
            && (level.getBlockState(north).`is`(block)
                    || level.getBlockState(south).`is`(block))
        ) {
            speed /= 2.0f
        } else {
            //判定斜角
            if (level.getBlockState(west.north()).`is`(block)
                || level.getBlockState(east.north()).`is`(block)
                || level.getBlockState(east.south()).`is`(block)
                || level.getBlockState(west.south()).`is`(block)
            ) {
                speed /= 2.0f
            }
        }
        return speed
    }

    override fun createBlockStateDefinition(builder: StateDefinition.Builder<Block?, BlockState?>) {
        builder.add(AGE)
    }

    override fun isRandomlyTicking(state: BlockState): Boolean = getAge(state) < MAX_AGE

    override fun isValidSugarTarget(
        level: LevelReader,
        pos: BlockPos,
        state: BlockState,
        isClient: Boolean,
    ): Boolean = getAge(state) < MAX_AGE

    override fun isSugarSuccess(
        level: ServerLevel,
        rand: RandomSource,
        pos: BlockPos,
        state: BlockState,
    ): Boolean = rand.nextFloat() < 0.8f

    override fun performSugar(
        level: ServerLevel,
        rand: RandomSource,
        pos: BlockPos,
        state: BlockState,
    ) {
        val age = min(getAge(state) + Random.nextInt(2, 5), MAX_AGE)
        level.setBlock(pos, state.setValue(AGE, age), 2)
    }
}